OpenRA.Game/Network/UPnP.cs

```diff
@@ -10,127 +10,77 @@
-
-				foreach (var mp in natDevice.GetAllMappings())
-					Log.Write("server", "Existing port mapping: protocol={0}, public={1}, private={2}",
-						mp.Protocol, mp.PublicPort, mp.PrivatePort);
-			}
-			catch (Exception e)
-			{
-				Log.Write("server", "Can't fetch information from NAT device: {0}", e);
+			NatDiscoverer.TraceSource.Switch.Level = SourceLevels.Verbose;
+			Log.AddChannel("nat", "nat.log");
+			var logChannel = Log.Channel("nat");
+			NatDiscoverer.TraceSource.Listeners.Add(new TextWriterTraceListener(logChannel.Writer));
 
-				Game.Settings.Server.NatDeviceAvailable = false;
-				Game.Settings.Server.AllowPortForward = false;
-			}
-		}
-
-		public static void ForwardPort(int lifetime)
-		{
-			try
-			{
-				var mapping = new Mapping(Protocol.Tcp, Game.Settings.Server.ExternalPort, Game.Settings.Server.ListenPort, lifetime);
-				natDevice.CreatePortMap(mapping);
-				Log.Write("server", "Create port mapping: protocol = {0}, public = {1}, private = {2}, lifetime = {3} s",
-					mapping.Protocol, mapping.PublicPort, mapping.PrivatePort, mapping.Lifetime);
-			}
-			catch (MappingException e)
+			var natDiscoverer = new NatDiscoverer();
+			var token = new CancellationTokenSource(timeout);
+			natDevices = await natDiscoverer.DiscoverDevicesAsync(PortMapper.Upnp, token);
+			foreach (var natDevice in natDevices)
 			{
-				if (e.ErrorCode == 725 && lifetime != 0)
+				try
 				{
-					Log.Write("server", "NAT device answered with OnlyPermanentLeasesSupported. Retrying...");
-					ForwardPort(0);
+					ExternalIP = natDevice.GetExternalIPAsync().Result;
 				}
-				else
+				catch (Exception e)
 				{
-					Log.Write("server", "Can not forward ports via UPnP: {0}", e);
-					Game.Settings.Server.AllowPortForward = false;
+					Console.WriteLine("Getting the external IP from NAT device failed: {0}", e.Message);
+					Log.Write("nat", e.StackTrace);
 				}
 			}
 		}
 
-		public static void RemovePortforward()
+		public static async void ForwardPort(int listen, int external)
 		{
-			try
-			{
-				var mapping = new Mapping(Protocol.Tcp, Game.Settings.Server.ExternalPort, Game.Settings.Server.ListenPort);
-				natDevice.DeletePortMap(mapping);
-				Log.Write("server", "Remove port mapping: protocol = {0}, public = {1}, private = {2}, expiration = {3}",
-					mapping.Protocol, mapping.PublicPort, mapping.PrivatePort, mapping.Expiration);
-			}
-			catch (Exception e)
+			mapping = new Mapping(Protocol.Tcp, listen, external, "OpenRA");
+			foreach (var natDevice in natDevices)
 			{
-				Log.Write("server", "Can not remove UPnP portforwarding rules: {0}", e);
-				Game.Settings.Server.AllowPortForward = false;
+				try
+				{
+					await natDevice.CreatePortMapAsync(mapping);
+				}
+				catch (Exception e)
+				{
+					Console.WriteLine("Port forwarding failed: {0}", e.Message);
+					Log.Write("nat", e.StackTrace);
+				}
 			}
 		}
 
-		public static IPAddress GetExternalIP()
+		public static async void RemovePortforward()
```

Comment:
Also, you should make the caller wait on this task to complete with either `await` or `.Wait()`.